src/stm32_generic/i2c.cpp

```diff
@@ -0,0 +1,548 @@
+    }
+    if (m_data_in.size() == 2 && m_reciever) {
+      bit_modify(i2c_reg->cr1).clear(i2c_cr1::ack_enable).set(i2c_cr1::pos);
+    }
+    [[maybe_unused]] auto addr = i2c_reg->sr2;
+    if (m_data_in.size() == 2 && m_reciever) {
+      while (!bit_extract<i2c_sr1::byte_transfered_finish>(status)) {
+        continue;
+      }
+      m_data_in[1] = data;
+      m_data_in[0] = data;
+    }
+
+    return;
+  }
+
+  if (bit_extract<i2c_sr1::tx_empty>(status)) {
+    if (m_data_out.empty()) {
+      // wait for transaction to finish
+      while (bit_extract<i2c_sr1::byte_transfered_finish>(status)) {
+        continue;
+      }
+      if (!m_data_in.empty()) {
+        bit_modify(i2c_reg->cr1).set(i2c_cr1::start);
+      } else {
+        bit_modify(i2c_reg->cr1).clear(i2c_cr1::ack_enable);
+        bit_modify(i2c_reg->cr1).set(i2c_cr1::stop);
+        m_busy = false;
+      }
+    } else {
+      data = m_data_out[0];
+      m_data_out = m_data_out.subspan(1);
+    }
+    return;
+  }
+
+  if (bit_extract<i2c_sr1::rx_not_empty>(status)) {
+    switch (m_data_in.size()) {
+      case 2: {
+        while (bit_extract<i2c_sr1::byte_transfered_finish>(status)) {
+          continue;
+        }
+        bit_modify(i2c_reg->cr1).clear(i2c_cr1::ack_enable);
+        m_data_in[0] = data;
+
+        while (bit_extract<i2c_sr1::byte_transfered_finish>(status)) {
+          continue;
+        }
+        bit_modify(i2c_reg->cr1).set(i2c_cr1::stop);
+        m_data_in[1] = data;
+        m_data_in[2] = data;
+        m_busy = false;
+        break;
+      }
+
+      case 1: {
+        while (bit_extract<i2c_sr1::byte_transfered_finish>(status)) {
+          continue;
+        }
+        m_data_in[0] = data;
+        bit_modify(i2c_reg->cr1).set(i2c_cr1::stop);
+        m_busy = false;
+        break;
+      }
+
+      case 0: {
+        bit_modify(i2c_reg->cr1).set(i2c_cr1::stop).clear(i2c_cr1::ack_enable);
+        [[maybe_unused]] auto a = data;
+        break;
+      }
+
+      default: {
+        m_data_in[0] = data;
+        m_data_in = m_data_in.subspan(1);
+      }
+    }
+  }
+}
+
+void i2c::handle_i2c_error() noexcept
+{
+  auto i2c_reg = get_i2c_reg(m_i2c);
+  bool soft_reset = false;
+  auto& status = i2c_reg->sr1;
+
+  if (bit_extract<i2c_sr1::bus_error>(status)) {
+    m_status = error_state::io_error;
+    bit_modify(i2c_reg->sr1).clear<i2c_sr1::bus_error>();
+    soft_reset = true;
+  }
+
+  if (bit_extract<i2c_sr1::ack_failure>(status)) {
+    m_status = error_state::no_such_device;
+    bit_modify(i2c_reg->sr1).clear<i2c_sr1::ack_failure>();
+    soft_reset = true;
+  }
+
+  if (bit_extract<i2c_sr1::timeout_error>(status)) {
+    /// we don't support stm32's timeout function, but have an external io
+    /// waiter
+    bit_modify(i2c_reg->sr1).clear(i2c_sr1::timeout_error);
+  }
+
+  if (bit_extract<i2c_sr1::arbitration_lost>(status)) {
+
+    m_status = error_state::arbitration_lost;
+    soft_reset = true;
+  }
```

Comment:
The order should be least to most severe:

1. arbitration_lost (and in this case, we don't throw anything, we should simply retry again
2. timeout_error
3. ack_failure
4. bus_error

The way your code is written currently, if you have both ack failure and bus error at the same time, we'd mask the io_error, which is way worse, for the no_such_device error.